Allow CustomTrainer to run a Python script directly via python_file

jskswamy opened the pull request:
**What this PR does / why we need it**:

This PR adds support for running a Python script directly in `CustomTrainer` by specifying a `python_file` argument. If `python_file` is set, the job will run the specified script as the main process (`python myscript.py`) instead of requiring a function. This is mutually exclusive with `func`; an error is raised if both or neither are set. The original function-based usage is unchanged and fully backward compatible.

- Simplifies migration from script-based workflows (e.g., Slurm, bash, or direct Kubernetes Jobs) to Kubeflow Trainer.
- Matches user expectations for direct script execution (`python myscript.py`).
- Avoids the indirection and complexity of function serialization and wrapper scripts.
- Ensures the script runs as the main process, improving signal handling and debugging.

**Which issue(s) this PR fixes**:

Fixes #47

**Summary of changes:**
- `CustomTrainer` now accepts an optional `python_file` argument.
- If `python_file` is set, the SDK sets the container entrypoint to `["python", python_file]`.
- Mutually exclusive with `func`; validation is added.
- No changes to existing function-based usage.

---

**Additional context:**
See [kubeflow/sdk#47](https://github.com/kubeflow/sdk/issues/47) for the original feature request and motivation.

eoinfennessy commented:
I'm not sure about this approach as it assumes the training script to run is available within the training runtime image. This would require users to rebuild training images each time a change is made to a training script.

Having said that, I think it would be a big UX improvement to allow users to provide a training script to run instead of a train function, but this script should come from the user's workspace instead of being baked into the training runtime image. This is related to issue #48, which describes making a snapshot of a user's workspaces available in training pods so that [multiple] Python files can be used in training jobs.

andreyvelich reviewed (commented):
Thank you for this @jskswamy!
I am wondering what are the benefits of such API compare to just this:


```python
from custom_script import run_pytorch

TrainerClient().train(
  trainer=CustomTrainer(
    func=run_pytorch,
    num_nodes=5
   )
)
```

If the TrainJob contains users' workspace, they should be able to execute the above script.
cc @shravan-achar

andreyvelich commented:
@jskswamy A few questions:
1. How you can ensure that Python file is inside the TrainJob workload ?
2. As you can see, entrypoint depends on [runtime Trainer](https://github.com/kubeflow/sdk/blob/main/kubeflow/trainer/constants/constants.py#L164). It could be `torchrun`, `mpirun`, or Python in case of PlainML.

jskswamy commented:
@andreyvelich  Thanks for pointing out the lack of framework-aware command support. I have added the following support now.

The `python_file` feature now uses the runtime's framework-specific commands instead of hardcoded `["python"]`:

- **PyTorch runtime**: Uses `torchrun train.py --args` 
- **MPI runtime**: Uses `mpirun python train.py --args`
- **PlainML runtime**: Uses `python train.py --args`

On the Python file availability, we are making an assumption that the Python file will be made available by the consumer through:

1. **Embedded in container image** - Scripts baked into the Docker image
2. **ConfigMaps** - Scripts mounted as ConfigMaps  
3. **Volume mounts** - Scripts mounted via PVC or other volumes
4. **Init containers** - Scripts downloaded by init containers

This design assumption provides flexibility for different deployment patterns while keeping the `python_file` feature simple and focused on direct script execution.

kramaranya reviewed (commented):
Thank you @jskswamy!

Have you considered how this affects the plugin architecture when someone writes a custom runtime?

kramaranya commented:
Hey @jskswamy! Did you have a chance to check the latest comments?

jskswamy commented:
I updated the original approach proposed in PR #49 to address the review concerns and make the feature more extensible and robust.

**What changed**
  - I replaced `python_file/python_args` with a new, explicit `CommandTrainer` that runs an arbitrary command with arguments under the runtime’s launcher.
  - The function path (`CustomTrainer`) remains intact, without special-casing or hardcoding.

**Why this fits the plugin architecture**
  - I reuse the runtime’s launcher template (torchrun/mpirun/python) and only replace the payload, so custom runtimes continue to control entrypoints and arguments.
  - No `command=["python"]` hardcoding; the launcher remains framework-aware. See discussion in [PR #49](https://github.com/kubeflow/sdk/pull/49).

**Distributed runtimes**
  - Torch uses `torchrun`, MPI uses `mpirun` (with `--user` pip installs), PlainML uses `python`.
  - This works for multi-node and device-aware training consistently.

**Shared features are preserved (no early return)**
  - Environment variables, resources, and pip installs are applied in the same builder path as functions.
  - The payload includes optional installs followed by the user command.

**File availability assumption**
  - The script/binaries must be present in the container (image/ConfigMap/volume/init). The workspace snapshot UX remains future work (issue #48 referenced in the PR thread).

**Validation strategy**
  - Consistent with `CustomTrainer`, I validate in the builder and error if `command` is missing/invalid.
  - I can add `__post_init__` later for earlier failure if needed, but builder-level checks match current patterns.

**Tests and docs**
  - Helper tests (plain/MPI) assert exact launcher shape and multiline payloads.
  - CRD builder tests cover env/resources/num_nodes and MPI `--user` installs.
  - Backend routing test verifies `CommandTrainer` dispatch.
  - README includes a minimal `CommandTrainer` usage example.

**Outcome**
  - Clear, framework-aware command execution without bypassing shared features or limiting plugin extensibility, and a clean separation from the function-based path.
  
@kramaranya , @andreyvelich kindly review the changes and let me know what you think about the new approach

jskswamy commented:
### Update on CommandTrainer behavior (pass-through + launcher-aware fallback)

- When `command` is provided and no `packages_to_install`: CRD uses `command`/`args` directly (no shell wrapper).
- When `packages_to_install` is set OR `command` is not provided: runtime launcher is reused via `get_command_using_user_command` (bash-wrapped), preserving installs and launcher selection (`torchrun`/`mpirun`/`python`).
- Mirrors BuiltinTrainer default launcher selection for the fallback path; explicit commands remain unchanged.
- Tests updated to cover both pass-through and launcher-wrapped paths.

Examples:
```python
# Direct (no wrapper)
trainer = types.CommandTrainer(command=["torchrun"], args=["--standalone", "train.py"])

# Install-aware (bash-wrapped)
trainer = types.CommandTrainer(
  command=["torchrun"],
  args=["--standalone", "train.py"],
  packages_to_install=["nemo", "deepspeed"],
)
```

andreyvelich commented:
This is resolved by: https://github.com/kubeflow/sdk/pull/127
/close
